mlx_graphs/data/data.py

```diff
@@ -47,19 +47,62 @@ def __init__(
             setattr(self, key, value)
 
     def __repr__(self):
-        return "%s(%s)" % (
-            type(self).__name__,
-            ", ".join(
-                "%s=%s" % (item[0], item[1].tolist())
-                for item in vars(self).items()
-                if item[1] is not None
-            ),
-        )
-
-    def to_dict(self):
+        strings = []
+        for k, v in vars(self).items():
+            if v is not None and not k.startswith("_"):
+                strings.append(f"{k}={v}")
+
+        return f"{type(self).__name__}({', '.join(strings)})"
+
+    def to_dict(self) -> dict:
         """Converts the Data object to a dictionary.
 
         Returns:
-            dict: A dictionary representation of the Data object.
+            A dictionary with all attributes of the `GraphData` object.
         """
         return {k: v for k, v in self.__dict__.items() if v is not None}
+
+    @property
+    def num_nodes(self) -> int:
+        return self.node_features.shape[0]
```

Comment:
`self.node_features` may be None

```suggestion
    @property
    def num_nodes(self) -> Optional[int]:
        if self.node_features:
            return self.node_features.shape[0]
        return None
```

Reply:
The None node_features is unexpected behavior for the moment, we didn't handle this case in messsage passing yet. 
I think I will have to do other modifications in the project to handle this case properly. 
Are node features also optional in PyG?

Reply:
Yes they are - see [here](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/data/data.html#Data). In general all attributes are optional so we'll have to handle that

Reply:
We can also do that in another PR btw.